soft_uart/example/main/soft_uart_main.c

```diff
@@ -0,0 +1,102 @@
+/*
+ * SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: CC0-1.0
+ */
+#include <stdio.h>
+#include <string.h>
+#include "sdkconfig.h"
+
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+
+#include "esp_log.h"
+#include "esp_check.h"
+#include "rmt_uart.h"
+
+#define READ_BUFFER_SIZE  1024
+#define EXAMPLE_TX_PIN  1
+#define EXAMPLE_RX_PIN  0
+
+const char *TAG = "rmt_uart_example";
+
+
+IRAM_ATTR static bool rmt_uart_tx_event_cbs(rmt_uart_device_handle_t tx_unit, const rmt_uart_tx_done_event_data_t *edata, void *user_ctx)
+{
+    BaseType_t high_task_wakeup = pdFALSE;
+    TaskHandle_t task = (TaskHandle_t)user_ctx;
+    vTaskNotifyGiveFromISR(task, &high_task_wakeup);
+    return high_task_wakeup == pdTRUE;
+}
+
+IRAM_ATTR static bool rmt_uart_rx_event_cbs(rmt_uart_device_handle_t rx_unit, const rmt_uart_rx_done_event_data_t *edata, void *user_ctx)
+{
+    BaseType_t high_task_wakeup = pdFALSE;
+    QueueHandle_t receive_queue = (QueueHandle_t)user_ctx;
+    xQueueSendFromISR(receive_queue, edata, &high_task_wakeup);
+    return high_task_wakeup == pdTRUE;
+}
+
+void uart_read_task(void *pvParameters)
+{
+    rmt_uart_device_handle_t uart_device = (rmt_uart_device_handle_t)pvParameters;
+    QueueHandle_t receive_queue = xQueueCreate(10, sizeof(rmt_uart_rx_done_event_data_t));   // queue for receive rmt rx done event
+    rmt_uart_event_rx_callbacks_t uart_rx_cbs = {
+        .on_rx_trans_done = rmt_uart_rx_event_cbs,
+    };
+    ESP_ERROR_CHECK(rmt_uart_register_rx_event_callbacks(uart_device, &uart_rx_cbs, receive_queue));
+    rmt_uart_rx_done_event_data_t rmt_rx_evt_data;
+    char read_buffer[READ_BUFFER_SIZE] = { 0 };
+    while (1) {
+        /* Read from the UART */
+        ESP_ERROR_CHECK(rmt_uart_receive(uart_device, (uint8_t *)read_buffer, sizeof(read_buffer)));
+        if (xQueueReceive(receive_queue, &rmt_rx_evt_data, portMAX_DELAY)) {
+            int read_len = rmt_uart_decode_data(uart_device, &rmt_rx_evt_data, (uint8_t *)read_buffer, sizeof(read_buffer), true);
```

Comment:
because we will call the `rmt_uart_receive` and `rmt_uart_decode_data` both from the task, can we merge them into a signle function?